logstash-plugins/usagetype-plugin/src/main/java/org/reactome/logstash/plugins/usagetype/util/RangeTree.java

```diff
@@ -0,0 +1,66 @@
+package org.reactome.logstash.plugins.usagetype.util;
+
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.TreeMap;
+
+public class RangeTree<N extends Number & Comparable<N>> extends TreeMap<Range<N>, String>
+{
+	private static final long serialVersionUID = 720362968233917584L;
+
+	public RangeTree()
+	{
+		super(new Comparator<Range<N>>()
+		{
+			@Override
+			public int compare(Range<N> o1, Range<N> o2)
+			{
+				return o1.compareTo(o2);
+			}
+		});
+	}
+
+	/**
+	 * Searches for a node whose key is a range which contains value. If there is more than one range that could contain value,
+	 * the first range encountered will be returned, it should be the range with the lowest starting value.
+	 * @param value The value to search for.
+	 * @return A Range object, or NULL if no Range could contain value.
+	 */
+	public Range<N> findRangeForValue(N value)
```

Comment:
I think the `N` type needs a param annotation for the JavaDoc to be complete

Reply:
If I add Javadoc at the class-level, isn't it a bit redundant to add it here as well?

Reply:
I think it might be recommended by JavaDoc.  I have compiled a JavaDoc jar where it will warn if the generic placeholder is not given a param annotation.  I'm not thrilled with the redundnacy either, but it does give context for the method parameter without having to implicitly know that you should look to the class docs.

From https://idratherbewriting.com/java-javadoc-tags/:


"@param tags must be provided for every parameter in a method or constructor. Failure to do so will create an error and warning when you render Javadoc.

Note that usually classes don’t have parameters. There is one exception: Generics. Generic classes are classes that work with different type of objects. The object is specified as a parameter in the class in diamond brackets: <>. Although the Javadoc guidance from Oracle doesn’t mention them, you can add a @param tag for a generic class to note the parameters for the generic class. See this StackOverflow post for details. Here’s an example from that page:"

Reply:
Recommended is not required... I may do this if I have time, but I think the COVID19 stuff takes priority at the moment.

Type parameters are not "normal" parameters - the word parameter is used, but they are not a value like other parameters, they are a description of the type of value something else will be, right? So I'm not sure they fit in with that "@param tags must be provided for every parameter in a method...". ...And now I'm starting to think that the fact that it would just be a copy-paste of a Javadoc comment somewhere else means maybe it _shouldn't_ be done...